conans/client/toolchain/cmake.py

```diff
@@ -0,0 +1,300 @@
+
+    @property
+    def configuration_types(self):
+        # Reverse index for the configuration_types variables
+        ret = defaultdict(list)
+        for conf, definitions in self._configuration_types.items():
+            for k, v in definitions.items():
+                ret[k].append((conf, v))
+        return ret
+
+
+class CMakeToolchain(object):
+    filename = "conan_toolchain.cmake"
+
+    _conan_set_libcxx = textwrap.dedent("""
+        macro(conan_set_libcxx)
+            if(DEFINED CONAN_LIBCXX)
+                conan_message(STATUS "Conan: C++ stdlib: ${CONAN_LIBCXX}")
+                if(CONAN_COMPILER STREQUAL "clang" OR CONAN_COMPILER STREQUAL "apple-clang")
+                    if(CONAN_LIBCXX STREQUAL "libstdc++" OR CONAN_LIBCXX STREQUAL "libstdc++11" )
+                        set(CONAN_CXX_FLAGS "${CONAN_CXX_FLAGS} -stdlib=libstdc++")
+                    elseif(CONAN_LIBCXX STREQUAL "libc++")
+                        set(CONAN_CXX_FLAGS "${CONAN_CXX_FLAGS} -stdlib=libc++")
+                    endif()
+                endif()
+                if(CONAN_COMPILER STREQUAL "sun-cc")
+                    if(CONAN_LIBCXX STREQUAL "libCstd")
+                        set(CONAN_CXX_FLAGS "${CONAN_CXX_FLAGS} -library=Cstd")
+                    elseif(CONAN_LIBCXX STREQUAL "libstdcxx")
+                        set(CONAN_CXX_FLAGS "${CONAN_CXX_FLAGS} -library=stdcxx4")
+                    elseif(CONAN_LIBCXX STREQUAL "libstlport")
+                        set(CONAN_CXX_FLAGS "${CONAN_CXX_FLAGS} -library=stlport4")
+                    elseif(CONAN_LIBCXX STREQUAL "libstdc++")
+                        set(CONAN_CXX_FLAGS "${CONAN_CXX_FLAGS} -library=stdcpp")
+                    endif()
+                endif()
+                if(CONAN_LIBCXX STREQUAL "libstdc++11")
+                    add_definitions(-D_GLIBCXX_USE_CXX11_ABI=1)
+                elseif(CONAN_LIBCXX STREQUAL "libstdc++")
+                    add_definitions(-D_GLIBCXX_USE_CXX11_ABI=0)
+                endif()
+            endif()
+        endmacro()
+    """)
+
+    _template_toolchain = textwrap.dedent("""
+        # Conan generated toolchain file
+        cmake_minimum_required(VERSION 3.0)  # Needed for targets
+
+        # Avoid including toolchain file several times (bad if appending to variables like
+        #   CMAKE_CXX_FLAGS. See https://github.com/android/ndk/issues/323
+        if(CONAN_TOOLCHAIN_INCLUDED)
+          return()
+        endif()
+        set(CONAN_TOOLCHAIN_INCLUDED true)
+
+        ########### Utility macros and functions ###########
+        {{ cmake_macros_and_functions }}
+        ########### End of Utility macros and functions ###########
+
+        # Configure
+        {% if generator_platform %}
+        set(CMAKE_GENERATOR_PLATFORM "{{ generator_platform }}" CACHE STRING "" FORCE){% endif %}
+        {% if toolset %}set(CMAKE_GENERATOR_TOOLSET "{{ toolset }}" CACHE STRING "" FORCE){% endif%}
```

Comment:
just a minor thing, maybe have if and endif everywhere on their own lines, for styling and consistency?